lib/services/tunnelconn.go

```diff
@@ -0,0 +1,237 @@
+package services
+
+import (
+	"encoding/json"
+	"fmt"
+	"runtime/debug"
+	"strings"
+	"time"
+
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/utils"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+)
+
+// TunnelConnection is SSH reverse tunnel connection
+// established to reverse tunnel proxy
+type TunnelConnection interface {
+	// Resource provides common methods for resource objects
+	Resource
+	// GetClusterName returns name of the cluster
+	// this connection is for
+	GetClusterName() string
+	// GetProxyName returns the proxy name this connection is established to
+	GetProxyName() string
+	// GetLastHeartbeat returns time of the last heartbeat received from
+	// the tunnel over the connection
+	GetLastHeartbeat() time.Time
+	// SetLastHeartbeat sets last heartbeat time
+	SetLastHeartbeat(time.Time)
+	// Check checks tunnel for errors
+	Check() error
+	// CheckAndSetDefaults checks and set default values for any missing fields.
+	CheckAndSetDefaults() error
+	// String returns user friendly representation of this connection
+	String() string
+	// Clone returns a copy of this tunnel connection
+	Clone() TunnelConnection
+}
+
+// MustCreateTunnelConnection returns new connection from V2 spec or panics if
+// parameters are incorrect
+func MustCreateTunnelConnection(name string, spec TunnelConnectionSpecV2) TunnelConnection {
```

Comment:
Looks like this is only used in tests, we should move it tests.

Reply:
tests are first-class users of the code, so it should be OK to keep it here. Just like `MustParseRegexp` in `regexp` standard package.